### PR TITLE
Don't suggest awaiting in closure patterns

### LINKED ISSUE #126903: Invalid suggestion to .await on a function argument pattern
### Code

```Rust
async fn do_async() {}

fn main() {
    Some(do_async()).map(|()| {});
}
```

### Current output

```Shell
Compiling playground v0.0.1 (/playground)
error[E0308]: mismatched types
 --> src/main.rs:4:27
  |
4 |     Some(do_async()).map(|()| {});
  |                           ^^
  |                           |
  |                           expected future, found `()`
  |                           expected due to this
  |
  = note: expected opaque type `impl Future<Output = ()>`
               found unit type `()`
help: consider `await`ing on the `Future`
  |
4 |     Some(do_async()).map(|().await| {});
  |                             ++++++

For more information about this error, try `rustc --explain E0308`.
error: could not compile `playground` (bin "playground") due to 1 previous error
```

### Desired output

```Shell
Don't suggest awaiting.
```

### Rationale and extra context

_No response_

### Other cases

_No response_

### Rust Version

```Shell
Playground's stable rust (1.79.0)
```

### Anything else?

Found by 0xbadbeef_60104 on the official rust discord.

### PR DESCRIPTION
**Pull request by SparkyPotato**
Fixes #126903.

For
```rust
async fn do_async() {}

fn main() {
    Some(do_async()).map(|()| {});
}
```
the error is now
```rust
error[E0308]: mismatched types
 --> src/main.rs:4:27
  |
4 |     Some(do_async()).map(|()| {});
  |                           ^^
  |                           |
  |                           expected future, found `()`
  |                           expected due to this
  |
  = note: expected opaque type `impl Future<Output = ()>`
               found unit type `()`
```

Ideally, if `main` were to be `async`, it should be
```rs
error[E0308]: mismatched types
 --> src/main.rs:4:27
  |
4 |     Some(do_async()).map(|()| {});
  |                           ^^
  |                           |
  |                           expected future, found `()`
  |                           expected due to this
  |
  = note: expected opaque type `impl Future<Output = ()>`
               found unit type `()`
help: consider `await`ing on the `Future`
  |
4 |     Some(do_async().await).map(|()| {});
  |                    ++++++
```
However, this would mean `FnCtx::check_pat_top` would have to be called with an `origin_expr` in `rustc_hir_typeck::check::check_fn`, and that expr would have to be somehow plumbed through `FnCtxt::check_expr_closure` and closure signature deduction. I'm willing to work on the plumbing but unsure how to start.

### CODE REVIEW

File: compiler/rustc_infer/src/infer/error_reporting/suggest.rs [hunk 1]
```diff
@@ -209,8 +209,10 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
             }
             (Some(ty), _) if self.same_type_modulo_infer(ty, exp_found.found) => match cause.code()
             {
-                ObligationCauseCode::Pattern { span: Some(then_span), .. } => {
-                    Some(ConsiderAddingAwait::FutureSugg { span: then_span.shrink_to_hi() })
+                ObligationCauseCode::Pattern { span: Some(then_span), origin_expr, .. } => {
+                    origin_expr.then_some(ConsiderAddingAwait::FutureSugg {
```
Comment (compiler-errors) [inline compiler/rustc_infer/src/infer/error_reporting/suggest.rs:213]: Is there any test that exercises the case when `origin_expr` is `Some`?

**Review by compiler-errors — APPROVED** [review]

**Comment by compiler-errors** [comment]
Please squash this into one commit

File: compiler/rustc_infer/src/infer/error_reporting/suggest.rs (hunk 1 above)
Comment (SparkyPotato) [reply compiler/rustc_infer/src/infer/error_reporting/suggest.rs:213]: yeah, most of `suggest-missing-await.rs` does

**Comment by compiler-errors** [comment]
thanks

@bors r+ rollup

**Pull request merged**